### PR TITLE
feat: course news feed

### PR DESCRIPTION
**Pull request by Chifez**
## What does this PR do?

Created the UI of the announcement page 

Fixes # (issue)

https://github.com/rotimi-best/classroomio/assets/69228399/1a9d4c0f-88ae-4957-8ff0-0b521cd6ff74

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] Enhancement (small improvements)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change adds a new database migration
- [ ] This change requires a documentation update

## How should this be tested?

- Ran it on my local host

### Required

- [x] Filled out the "How to test" section in this PR
- [ ] Self-reviewed my own code
- [ ] Commented on my code in hard-to-understand bits
- [ ] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [ ] Removed all `console.logs`
- [ ] Merged the latest changes from main onto my branch with `git pull origin main`
- [ ] My changes don't cause any responsiveness issues

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [ ] Updated the ClassroomIO Docs if changes were necessary

### CODE REVIEW

**Comment by rotimi-best** [comment]
@Chifez I have watched your demo. The frontend of this looks really good, great job on it.

I will make this PR draft for now because:
1. I need to improve our notification functionality before we ship this one.
2. I want you to implement the backend of this task.
3. You should complete the blog task first https://github.com/rotimi-best/classroomio/issues/139

**Comment by Chifez** [comment]
> @Chifez I have watched your demo. The frontend of this looks really good, great job on it.
> 
> Please make this PR draft for now because:
> 
> 1. I need to improve our notification functionality before we ship this one.
> 2. I want you to implement the backend of this task.
> 3. You should complete the blog task first [Create classroomio blog #139](https://github.com/rotimi-best/classroomio/issues/139)

alright noted

File: apps/dashboard/src/lib/components/Course/index.svelte [hunk 1]
```diff
@@ -54,3 +54,162 @@
     <HtmlRender className="p-2" content={$course.overview || ''} />
   {/if}
 </PageBody>
+
+<!-- <script>
+  // @ts-nocheck
+
+  import PageNav from '../PageNav/index.svelte';
+  import PrimaryButton from '$lib/components/PrimaryButton/index.svelte';
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/components/Course/index.svelte]: What is this for?

File: apps/dashboard/src/lib/utils/functions/formatTime.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+export default function formatTime(utcTimestamp) {
+  const date = new Date(utcTimestamp);
+  let hours = date.getUTCHours();
+  const minutes = date.getUTCMinutes();
+  const ampm = hours >= 12 ? 'PM' : 'AM';
+  hours = hours % 12 || 12;
+  const formattedTime = `${hours}:${minutes.toString().padStart(2, '0')}${ampm}`;
+  return formattedTime;
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/functions/formatTime.js]: What's the goal of this function?

There is an implementation of how we use `dayjs`, let me know your use case and I'll properly direct you.

File: apps/dashboard/src/lib/utils/services/announcement/index.ts [hunk 3]
```diff
@@ -0,0 +1,93 @@
+import type { Announcement } from '$lib/utils/types/anouncement';
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/announcement/index.ts]: @Chifez I thought you mentioned you have changed everything from announcement to newsfeed?

This file has announcement, this line and the types file has announcement.

Please go over everything again.

File: apps/dashboard/src/lib/utils/services/announcement/index.ts [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    comment:course_newsfeed_comment(id,
+        created_at,
+        author_id,
+        content,
+        course_newsfeed_id)
+    `
+    )
+    .match({ course_id: courseId })
+    .order('created_at', { ascending: false });
+  //   const response: any = await supabase.from('course_announcement').select('*');
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/announcement/index.ts]: Please remove

File: apps/dashboard/src/lib/utils/services/announcement/index.ts [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    .from('course_newsfeed')
+    .select(
+      `
+    id,
+    created_at,
+    content,
+    course_id,
+    author_id,
+    reaction,
+    comment:course_newsfeed_comment(id,
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/announcement/index.ts]: please bring `id,` to a new line.

File: apps/dashboard/src/lib/utils/services/announcement/index.ts [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+  return { data, error };
+}
+
+export async function createAnnouncement(announcement: {
+  content: string;
+  author_id: string;
+  course_id: string;
+  reaction: any;
+}) {
+  console.log(announcement);
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/announcement/index.ts]: remove

File: apps/dashboard/src/lib/utils/services/announcement/index.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+    .from('course_newsfeed')
+    .insert({
+      content: announcement.content,
+      author_id: announcement.author_id,
+      course_id: announcement.course_id,
+      reaction: announcement.reaction
+    })
+    .select();
+
+  console.log('create', response);
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/announcement/index.ts]: remove

File: apps/dashboard/src/lib/utils/services/announcement/index.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+  const response = await supabase
+    .from('course_newsfeed_comment')
+    .insert({
+      content: comment.content,
+      author_id: comment.author_id,
+      course_newsfeed_id: comment.course_newsfeed_id
+    })
+    .select();
+
+  console.log('create', response);
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/announcement/index.ts]: remove

File: apps/dashboard/src/lib/utils/services/announcement/index.ts [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+
+  return { response };
+}
+
+export async function deleteNewsFeedComment(commentId: string) {
+  const response = await supabase.from('course_newsfeed_comment').delete().match({ id: commentId });
+  console.log('deletecomment', response);
```
Comment (rotimi-best) [inline apps/dashboard/src/lib/utils/services/announcement/index.ts]: remove

**Review by rotimi-best — CHANGES_REQUESTED** [review]
There are still so many console.logs, please review and remove them

File: apps/dashboard/src/lib/utils/functions/formatTime.js (hunk 2 above)
Comment (Chifez) [reply apps/dashboard/src/lib/utils/functions/formatTime.js]: the created_at timestamp from the database is converted to a 12hr time format here,i couldnt find any pre existing function in the utils so i wrote this

Comment (Chifez) [reply apps/dashboard/src/lib/utils/functions/formatTime.js]: right now it just indicates the formatted created _at time like 9:33 AM but something i intend to do is such that it displays yesterday or the full date it was sent like 11 feb 12:00 AM, when the feed is an old feed

Comment (Chifez) [reply apps/dashboard/src/lib/utils/functions/formatTime.js]: the dayjs formatDate function in the utils file only returns today, yesterday

Comment (Chifez) [reply apps/dashboard/src/lib/utils/functions/formatTime.js]: i implemented it using dayjs,i read the docs and the format date code and figured how to do it,a dayjs inbuilt plugin called localizedFormat was used 
Screenshot (716)

i am trying to send a screenshot but its getting broken

Comment (Chifez) [reply apps/dashboard/src/lib/utils/functions/formatTime.js]: its probably because i am using airtel

Comment (rotimi-best) [reply apps/dashboard/src/lib/utils/functions/formatTime.js]: Use this function `calDateDiff`.

image

We already use it on the community question page and many other places.
image

Comment (Chifez) [reply apps/dashboard/src/lib/utils/functions/formatTime.js]: done

**Comment by rotimi-best** [comment]
# UX Feedbacks

## 1. Check 3 dot icon.

Instead of this
image

use this one
image

https://carbon-components-svelte.onrender.com/components/OverflowMenu

## 2. Implement Edit functionality

Add edit to the overflow menu and open a modal with the content prefilled.

image

## 3. When I save, the content losses it HTML formatting

This is what I entered in the editor:
image

This is what is displayed in the UI
image

Note: Look at how it was implemented on the Lesson editing and view mode

## 4 When I comment on a post, and reload avatar is broken
image

## 5. Snackbar notifications were not implemented
Show snackbar notification when the following happens

1. Create a new post
2. Delete a post
3. Edit a post
4. Commentted on a post

Also make sure to show error notifications as well when an error occurs in any of the above steps.

## 6. Show a warning modal on delete
Right now when you delete, it just deletes immediately. Please see how deleting a lesson works and follow something similar
image

**Comment by Chifez** [comment]
> Functionality for newsfeed is complete, however there are a couple of things that I need to put in place for us to deploy this feature.
> 
> # Checklist
> * [ ]  Implement email notification with https://defer.run/
> * [ ]  Write template functions for discord and telegram notification
> * [ ]  Write blog post announcing feature.

oh ok,  can i help with anything?

**Comment by rotimi-best** [comment]
> Functionality for newsfeed is complete, however there are a couple of things that I need to put in place for us to deploy this feature.
> 
> # Checklist
> * [ ]  Implement email notification
> * [ ]  Implement pinned comment
> * [ ]  Write template functions for discord and telegram notification
> * [ ]  Write blog post announcing feature.


## I just added pinned feed.

image

> oh ok, can i help with anything?

@Chifez Not really. I am doing some refactoring and adding some finishing touches to the UX and the release. 👍🏽

**Comment by rotimi-best** [comment]
Removed the overview on the sidebar
image

**Comment by rotimi-best** [comment]
> Functionality for newsfeed is complete, however there are a couple of things that I need to put in place for us to deploy this feature.
> 
> # Checklist
> * [ ]  Implement email notification
> * [x]  Implement pinned comment
> * [ ]  Write template functions for discord and telegram notification

I have removed the point of writing the blog post. I plan on making this part of an upcoming launch week.

We will launch 5 new features in 5 days and this will be one of them.

Once I am done with the notification part, I will merge it in and we will keep testing it before the launch week. The date for the launch has not been finalized yet.

**Review by rotimi-best — APPROVED** [review]
This work has come a long way. Great job @Chifez.

I would encourage you to go over the code to see the refactoring I made.

This is important so that before your next PR you will get better

**Pull request merged**